.github/workflows/ci.yaml

```diff
@@ -0,0 +1,112 @@
+
+      # License activation sometimes fails due to a problem connecting
+      # to Unity's licensing backend. Retry a few times if it fails.
+      - name: Activate license
+        run: |
+          $isSuccessful = $False
+          $attempts = 5
+
+          while ($attempts -gt 0) {
+            try {
+              unity -quit -batchmode -nographics -logFile - -serial ${{ secrets.UNITY_SERIAL }} -username ${{ secrets.UNITY_EMAIL }} -password ${{ secrets.UNITY_PASSWORD }} | Out-Default
+
+              $isSuccessful = $True
+              break
+            } catch {
+              Start-Sleep -Seconds 10
+            }
+
+            $attempts--
+          }
+
+          if (-not $isSuccessful) {
+            Write-Host "::error ::Failed to activate license after multiple attempts."
+            exit 1
+          }
+
+      - name: Build project
+        run: |
+          unity -quit -batchmode -nographics -logFile - -projectPath ${{ env.PROJECT_PATH }} -buildWindows64Player artifacts/build/game.exe | Out-Default
+
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@v1
+        with:
+          name: Build output
+          path: ${{ env.PROJECT_PATH }}/artifacts/build
+
+      # This command doesn't work with the `-quit` option, unlike all the other ones
+      - name: Run tests
+        run: |
+          unity -batchmode -nographics -logFile - -runTests -testPlatform StandaloneWindows64 -projectPath ${{ env.PROJECT_PATH }} -testResults artifacts/test/results.xml | Out-Default
+
+      - name: Upload test artifacts
+        uses: actions/upload-artifact@v1
+        with:
+          name: Test results
+          path: ${{ env.PROJECT_PATH }}/artifacts/test
+
+      # Running tests does not print anything useful to the console, nor does it fail
+      # the build, so we need to parse the results to figure out how things went.
+      - name: Parse test results
+        run: |
+          $testRun = Select-Xml -Path ${{ env.PROJECT_PATH }}/artifacts/test/results.xml -XPath "/test-run" | Select-Object -Last 1
+
+          Write-Output "Test run completed in $($testRun.Node.duration)."
+          Write-Output "-- Result: $($testRun.Node.result)"
+          Write-Output "-- Total: $($testRun.Node.total)"
+          Write-Output "-- Passed: $($testRun.Node.passed)"
+          Write-Output "-- Failed: $($testRun.Node.failed)"
+          Write-Output "-- Inconclusive: $($testRun.Node.inconclusive)"
+          Write-Output "-- Skipped: $($testRun.Node.skipped)"
+
+          if ($testRun.Node.failed -gt 0) {
+            Write-Output "::error ::Test run completed with $($testRun.Node.failed) failing test(s)."
+            exit 1
+          }
+
+      # Professional licenses are per-seat so we should always try to return them
+      - name: Return license
+        run: |
+          unity -quit -batchmode -nographics -logFile - -returnlicense | Out-Default
```

Comment:
Does this need a retry too?
I wonder if this starts failing how do we recover the license.
Maybe worth considering adding Sentry here (.net sdk) and raising an event when it happens so we can be notified and recover the license.

Reply:
The retry here will be more complicated because this step may fail if the activation step failed (since this one runs unconditionally). Also, to identify failure, we'll need to do some test parsing on the output.

Let's leave it like this for now and see what happens. My understanding is that the license should be freed up after some time.